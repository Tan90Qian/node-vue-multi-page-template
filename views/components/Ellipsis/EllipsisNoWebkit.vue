<template>
  <div ref="root" :class="cls">
    <div ref="content">
      <span ref="node">{{ childText }}</span>
      <div class="shadow" ref="shadowChildren">{{ children }}</div>
      <div class="shadow" ref="shadow">
        <span>{{ text }}</span>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: ['cls', 'children', 'lines'],
  data() {
    return {
      text: '',
      targetCount: 0,
    };
  },
  mounted() {
    if (this.$refs.node) {
      this.computeLine();
    }
  },
  watch: {
    lines(val, oldVal) {
      if (val !== oldVal) {
        this.computeLine();
      }
    },
  },
  methods: {
    computeLine() {
      const { lines } = this;
      const text = this.$refs.shadowChildren.innerText;
      const lineHeight = parseInt(getComputedStyle(this.$refs.root).lineHeight, 10);
      const targetHeight = lines * lineHeight;
      this.$refs.content.style.height = `${targetHeight}px`;
      const totalHeight = this.$refs.shadowChildren.offsetHeight;
      const shadowNode = this.$refs.shadow.firstChild;

      if (totalHeight <= targetHeight) {
        this.text = text;
        this.targetCount = text.length;
        return;
      }

      // bisection
      const len = text.length;
      const mid = Math.ceil(len / 2);

      const count = this.bisection(targetHeight, mid, 0, len, text, shadowNode);

      this.text = text;
      this.targetCount = count;
    },
    bisection(th, m, b, e, text, shadowNode) {
      const suffix = '...';
      let mid = m;
      let end = e;
      let begin = b;
      shadowNode.innerHTML = text.substring(0, mid) + suffix;
      let sh = shadowNode.offsetHeight;

      if (sh <= th) {
        shadowNode.innerHTML = text.substring(0, mid + 1) + suffix;
        sh = shadowNode.offsetHeight;
        if (sh > th || mid === begin) {
          return mid;
        } else {
          begin = mid;
          if (end - begin === 1) {
            mid = 1 + begin;
          } else {
            mid = Math.floor((end - begin) / 2) + begin;
          }
          return this.bisection(th, mid, begin, end, text, shadowNode);
        }
      } else {
        if (mid - 1 < 0) {
          return mid;
        }
        shadowNode.innerHTML = text.substring(0, mid - 1) + suffix;
        sh = shadowNode.offsetHeight;
        if (sh <= th) {
          return mid - 1;
        } else {
          end = mid;
          mid = Math.floor((end - begin) / 2) + begin;
          return this.bisection(th, mid, begin, end, text, shadowNode);
        }
      }
    },
  },
  computed: {
    childText() {
      const { text, targetCount } = this;
      return `${targetCount > 0 && text.substring(0, targetCount)}${
        targetCount > 0 && targetCount < text.length ? '...' : ''
      }`;
    },
  },
};
</script>

<style lang="scss" scoped></style>
