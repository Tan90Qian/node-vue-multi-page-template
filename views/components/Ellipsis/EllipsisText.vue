<template>
  <span v-if="!text"></span> <span v-else-if="ifRenderText">{{ text }}</span>
  <span v-else :title="text">{{ displayText }} {{ tail }}</span>
</template>

<script>
import { getStrFullLength, cutStrByFullLength } from './utilFnc.js';
export default {
  name: 'EllipsisText',
  props: ['text', 'length'],
  data() {
    return {
      tail: '...'
    };
  },
  mounted() {
    const { text } = this;
    if (typeof text !== 'string') {
      throw new Error('Ellipsis children must be string.');
    }
  },
  computed: {
    ifRenderText() {
      const { text, length } = this;
      const textLength = fullWidthRecognition ? getStrFullLength(text) : text.length;
      return textLength <= length || length < 0;
    },
    displayText() {
      const { text, length, tail } = this;
      if (length - tail.length <= 0) {
        return '';
      }
      return fullWidthRecognition ? cutStrByFullLength(text, length) : text.slice(0, length);
    }
  }
};
</script>
