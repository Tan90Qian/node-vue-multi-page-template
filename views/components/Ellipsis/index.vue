<template>
  <span v-if="!lines && !length" :class="cls"> <slot></slot> </span>
  <EllipsisText
    v-else-if="!lines"
    :class="cls"
    :length="length"
    :text="childrenText"
    :fullWidthRecognition="fullWidthRecognition"
  ></EllipsisText>
  <div
    v-else-if="isSupportLineClamp"
    :class="cls"
    :style="{
      '-webkit-line-clamp': lines,
      '-webkit-box-orient': 'vertical'
    }"
  >
    <slot></slot>
  </div>
  <EllipsisNoWebkit v-else :class="cls" :lines="lines" :children="childrenText"></EllipsisNoWebkit>
</template>

<script>
import EllipsisNoWebkit from './EllipsisNoWebkit';
import EllipsisText from './EllipsisText';
export default {
  name: 'Ellipsis',
  props: {
    lines: {
      type: Number
    },
    length: {
      type: Number
    }
  },
  components: {
    EllipsisNoWebkit,
    EllipsisText
  },
  computed: {
    isSupportLineClamp() {
      return document.body.style.webkitLineClamp !== undefined;
    },
    cls() {
      const { lines, isSupportLineClamp } = this;
      return {
        ellipsis: true,
        lines: lines && !isSupportLineClamp,
        lineClamp: lines && isSupportLineClamp
      };
    },
    childrenText() {
      const children = this.$slots.default;
      return children ? children[0].text : '';
    }
  }
};
</script>

<style lang="scss">
.ellipsis {
  display: inline-block;
  width: 100%;
  overflow: hidden;
  word-break: break-all;

  &.lines {
    position: relative;
    .shadow {
      position: relative;
      display: block;
      color: transparent;
      opacity: 0;
      z-index: -999;
    }
  }

  &.lineClamp {
    position: relative;
    display: -webkit-box;
    overflow: hidden;
    text-overflow: ellipsis;
  }
}
</style>
