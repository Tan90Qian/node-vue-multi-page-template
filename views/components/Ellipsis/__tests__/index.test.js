import { shallowMount } from '@vue/test-utils';

import Ellipsis from '../index';

describe('Ellipsis tests', () => {
  let children;

  beforeEach(() => {
    children = '一二，a,';
  });
  /* snapshot test */
  test('renders correctly', () => {
    const wrapper = shallowMount(Ellipsis, {
      slots: {
        default: [children]
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  /* render test */
  it('render span when absent lines and length', () => {
    const wrapper = shallowMount(Ellipsis, {
      slots: {
        default: [children]
      }
    });
    expect(wrapper.html()).toEqual(`<span class="ellipsis">${children}</span>`);
  });

  it('render empty span when children was absent', () => {
    const wrapper = shallowMount(Ellipsis);
    expect(wrapper.html()).toEqual('<span class="ellipsis"></span>');
  });

  it('render EllipsisText when lines was absent', () => {
    const wrapper = shallowMount(Ellipsis, {
      propsData: {
        length: 10
      }
    });
    expect(wrapper.html().toLowerCase()).toContain('EllipsisText'.toLowerCase());
  });

  it('render EllipsisText when lines was absent', () => {
    const wrapper = shallowMount(Ellipsis, {
      propsData: {
        length: 10
      }
    });
    expect(wrapper.html().toLowerCase()).toContain('EllipsisText'.toLowerCase());
  });
});
