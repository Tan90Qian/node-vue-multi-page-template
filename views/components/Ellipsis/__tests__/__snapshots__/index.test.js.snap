// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ellipsis tests renders correctly 1`] = `
<span
  class="ellipsis"
>
  一二，a,
</span>
`;
